### ISSUE TITLE
[Bug]: Overriding a member loses nullability analysis from base class

### ISSUE THREAD

**Issue by just-ero** [dotnet/roslyn#71740]
### Version Used

* Compiler version: '4.10.0-1.24058.1 (1c338c6a)'
* VSCode: 1.85.2, 8b3775030ed1a69b13e4f4c628c612102e30a681, x64
* C# extension: v2.15.30, C# Dev Kit extension: v1.3.4

### Steps to Reproduce

1. Open or create any C# project in Visual Studio Code.
2.
    ```cs
    class Class1 {
        public virtual string? Foo { get; }
    
        [MemberNotNull(nameof(Foo))]
        public void EnsureFooNotNull() {
            if (Foo is null) throw new();
        }
    }
    
    class Class2 : Class1 {
        public override string? Foo { get; }
    
        public string Bar() {
            EnsureFooNotNull();
            return Foo; // CS8603: Possible null reference return.
        }
    }
    ```

### Expected Behavior

No warning is shown.

### Actual Behavior

The compiler warns that `Foo` may be null when `EnsureFooNotNull()` already asserted that it is not.

**Comment by jaredpar** [comment]
@RikkiGibson, @jcouv can you comment on this? There isn't a complete spec for `MemberNotNull` so hard to say if this is / isn't expected. My intuition is that this sholud work as it's an `override` not a `new`.

**Comment by RikkiGibson** [comment]
I think that the handling of MemberNotNull should be enhanced to understand the semantics of 'override', by tying together the flow state of the base and override properties.

Consider a usage like the following: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACADAAT4CMAdACJQQDmMA9gM4JQDGzFAwoxgKYBBGBAA2AT2ZRmAbgCwAKEUA3CMmLtREZswBMxALzEY/AO7FuWnboAUASnkLN2vRQBCa+48X59ll2TEAN6KxGGkAMzEylDICKhipGSEAPzEAGKMjMHEdPwIMsTM+YUAvoqh4QDaALL8YMD8yAByjAjNqKKiNiJg/IwAZjaZjHZ2ALqVYfhR+AAsxACiMMyoyPwjre2d3XbBU+FhI4bEAEQQp46H5Qo3Pn5WesQgFo+BIQqHM8SMyk3IUAESVSGSyOTyBSKJWId0+4W+82IHmQ9n2cMOYWWq3WmzaHS6XgOhxGFAAKowAMoIAEwOheYgAegZxmypjUMFgdGI/AAHgAHfjsBD8DDEYCoBA/EzENnMYjrPmMOIiihE8LAbT8Cgk8lUml0hyM5m8gVC4gQGXszliiW5NoskxwYjNABKi3NkrQMDYfQON1KQA==)

```cs
using System.Diagnostics.CodeAnalysis;

var class2 = new Class2();
class2.Bar();

class Class1 {
    public virtual string? Foo { get; set; }

    [MemberNotNull(nameof(Foo))]
    public void EnsureFooNotNull() {
        Foo = "a";
    }
}

class Class2 : Class1 {
    public override string? Foo { get; set; }

    public void Bar() {
        EnsureFooNotNull();
        Foo.ToString(); // no warning expected but one was reported.
        base.Foo.ToString(); // no warning, NRE at runtime
    }
}
```

First, consider the unexpected warning on `Foo.ToString()`. Here I think we should make a change such that, when we try to get a slot for an override, we instead look up the least overridden member for it, and get a slot for that instead. This should remove the unexpected warning. Basically, override properties simply do not have their own slots. This is because most overridden property is used *anyway* when the property is accessed through an expression of type `Class1`, and in that context we need some slot to track the property.

Second, consider the missing warning and NRE on `base.Foo.ToString()`. I think it might be fine to just leave this hole in. I am pretty sure it resembles a required properties scenario where LDM decided it was fine to have the possibility of accessing the property through `base` before it is assigned.

We would also expect a change to constructor analysis as a result of this, for example:

```cs
class Base
{
    public virtual string Prop { get; set; }
    public Base() { Prop = "a"; }
}

class Derived
{
    public override string Prop { get; set; } // no warning to initialize Derived.Prop, because Base..ctor() actually initialized it.
}
```

**Comment by jaredpar** [comment]
> Second, consider the missing warning and NRE on base.Foo.ToString(). I think it might be fine to just leave this hole in. I am pretty sure it resembles a required properties scenario where LDM decided it was fine to have the possibility of accessing the property through base before it is assigned.

I agree. `base.` access is probably not worth digging down on here. Lots of ways it can trip up the system and there is reasonable expectation that within an inheritance graph the types understand the invariants they need to maintain. 

>  Here I think we should make a change such that, when we try to get a slot for an override, we instead look up the least overridden member for it, and get a slot for that instead

Any thoughts on how `[MemberNotNull]` should be treated when its applied to a `virtual` method?

Note: before moving forward on this think we need to spec the behavior out in the nullable ref type spec.

**Comment by RikkiGibson** [comment]
> Any thoughts on how `[MemberNotNull]` should be treated when its applied to a `virtual` method?

Ideally we would require that overrides of such a method repeat the same `[MemberNotNull]` in order to fulfill the promise the base method is making. But, I feel like it's not worth the design effort right now to introduce the ability for `[MemberNotNull]` to refer to members of a base type. So, I feel comfortable with it being a "hole"--if overrides neglect to ensure the indicated members are not null, then unexpected NREs may occur at the use site.
